session.go

```diff
@@ -1421,11 +1454,25 @@ func (s *Session) NewBatch(typ BatchType) *Batch {
 		Cons:             s.cons,
 		defaultTimestamp: s.cfg.DefaultTimestamp,
 		keyspace:         s.cfg.Keyspace,
+		metrics:          make(map[string]*queryMetrics),
 	}
 	s.mu.RUnlock()
 	return batch
 }
 
+func (b *Batch) getHostMetrics(host *HostInfo) *queryMetrics {
+	b.mu.Lock()
```

Comment:
The same suggestion I had above with changing this machinery applies to the `Batch` as well. Shame that it's duplicated. This screams separate manager object that holds query attempt metrics as well as latencies and total attempts. It would also include the mutex. Sorry for bringing it up. 😇

Reply:
I totally agree :-) I've almost implemented it on Friday, just was too sleepy to make sure the tests pass. Hopefully it will resolve all the gripes with the current implementation.